### ISSUE TITLE
feat: ONNX import — parse ONNX into CompiledInferencePlan

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#169]
## Motivation

Every major ML framework (PyTorch, TF, JAX) can **export** to ONNX. **Nobody imports ONNX and makes it fast.** ONNX Runtime does, but it's a separate runtime, not a first-class library.

If AiDotNet becomes the fastest way to run any ONNX model, we pick up an entire market segment. Users with PyTorch-trained models get our compile, fusion, CUDA Graph, AOT serialization, and autotune benefits — without any porting work.

## Proposed API

```csharp
public static class OnnxImporter
{
    /// <summary>
    /// Parses an ONNX model from a stream and produces a compiled inference plan
    /// ready for Execute(). All ONNX operators are translated to native tensor
    /// engine ops; ops that have no equivalent are either polyfilled via composed
    /// engine ops OR reported via UnsupportedOperators in the result.
    /// </summary>
    public static OnnxImportResult<T> Import<T>(
        Stream onnxModel,
        OnnxImportOptions? options = null)
        where T : unmanaged;
}

public sealed class OnnxImportResult<T>
{
    public ICompiledPlan<T>? Plan { get; }
    public ParameterBuffer<T>? Parameters { get; }
    public IReadOnlyList<string> UnsupportedOperators { get; }
    public IReadOnlyDictionary<string, ShapeProfile> NamedInputs { get; }
    public IReadOnlyDictionary<string, ShapeProfile> NamedOutputs { get; }
}
```

## Scope — operator coverage

**Phase 1 (minimum viable):** enough to import all image classification + transformer models from Hugging Face `optimum`:
- MatMul, Gemm, Add, Mul, Relu, Sigmoid, Tanh, Softmax, GELU, LayerNormalization, BatchNormalization
- Conv, ConvTranspose (2D + 3D), MaxPool, AveragePool, GlobalAveragePool
- Reshape, Transpose, Slice, Concat, Split, Gather, Squeeze, Unsqueeze
- Attention (from `com.microsoft.Attention` extension) → maps to our FusedAttention

**Phase 2:** quantized ops (QLinearConv, QLinearMatMul), recurrent ops (LSTM, GRU)

**Phase 3:** custom ops via user-provided polyfills

## Integration (AiDotNet)

```csharp
// User brings any ONNX model
var result = await AiModelBuilder<float, Tensor<float>, Tensor<float>>
    .FromOnnx("path/to/model.onnx")
    .ConfigureJitCompilation()
    .ConfigureAutotune()
    .BuildAsync();

// Now runs through our compile+autotune+AOT pipeline
var prediction = result.Predict(image);
```

## Acceptance criteria

- Imports ResNet-50 + BERT-base + ViT-B/16 from ONNX model zoo
- Numerical accuracy within 1e-4 of ONNX Runtime outputs on 100 validation samples
- Benchmark: compiled AiDotNet matches or beats ONNX Runtime on CPU latency for same models (the win should come from our fusion + CUDA Graph + autotune stack)
- Unsupported operators are **named and reported** rather than silently replaced

## Reference

Plan: `lucky-waddling-knuth` Part C PR 9. Repositions AiDotNet as an ONNX runtime competitor, not just a C# ML library.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

Nothing. No ONNX parser. `CompiledInferencePlan` exists (target of the import).

## 2. Scope — full inventory

### API
```csharp
public static class OnnxImporter
{
    public static OnnxImportResult<T> Import<T>(
        Stream onnxModel,
        OnnxImportOptions? options = null
    ) where T : unmanaged;
}

public sealed class OnnxImportResult<T>
{
    public ICompiledPlan<T>? Plan { get; }
    public ParameterBuffer<T>? Parameters { get; }
    public IReadOnlyList<string> UnsupportedOperators { get; }
    public IReadOnlyDictionary<string, ShapeProfile> NamedInputs { get; }
    public IReadOnlyDictionary<string, ShapeProfile> NamedOutputs { get; }
}
```

### Phase 1 (MVP — enough for HF optimum image + transformer models)
```
MatMul, Gemm, Add, Mul, Relu, Sigmoid, Tanh, Softmax, GELU,
LayerNormalization, BatchNormalization
Conv (2D + 3D), ConvTranspose, MaxPool, AveragePool, GlobalAveragePool
Reshape, Transpose, Slice, Concat, Split, Gather, Squeeze, Unsqueeze
Attention (from com.microsoft.Attention extension) → maps to our FusedAttention
```

### Phase 2 (quantized + recurrent)
```
QLinearConv, QLinearMatMul, DequantizeLinear, QuantizeLinear
LSTM, GRU (maps to cuDNN RNN from #219)
```

### Phase 3 (custom ops via polyfills)
```
User-provided OnnxOpTranslator<T> that maps unknown ops to our primitives
```

**Total: ~35 ops Phase 1 + ~8 Phase 2 + extension point.**

## 3. Win vectors

1. **Goal: fastest way to run any ONNX model** — PyTorch/TF/JAX can export but don't import well; ONNX Runtime imports but is a separate runtime, not a library.
2. **Direct to `CompiledInferencePlan`**: imported model immediately benefits from our fusion passes (Spectral/Codec/Dataflow/Attention), `AutotuneCache`, CUDA Graph capture, AOT serialization.
3. **Sub-byte on import**: int1/2/3/4/NF4/FP4 quantized ONNX models load natively.
4. **Attention extension → FusedAttention**: direct mapping to FlashAttention-2.
5. **6-backend execution** of imported model — ONNX Runtime = CPU + CUDA + DirectML + (some others).
6. **Named & reported unsupported operators**: users get a clear list, not silent wrong behavior.
7. **Integration with `AiModelBuilder`** for the upstream AiDotNet facade.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| ResNet-50 import + execute | numerical accuracy within 1e-4 of ONNX Runtime outputs on 100 val samples |
| BERT-base import + execute | within 1e-4 |
| ViT-B/16 import + execute | within 1e-4 |
| CPU latency vs ONNX Runtime (same 3 models) | ≥ parity; wins via fusion + CUDA Graph + autotune |
| Unsupported op reporting | named + reported; never silently replaced |
| Phase 2 quantized | QLinearConv/QLinearMatMul load + execute correctly |
| Phase 2 LSTM/GRU | maps to cuDNN RNN via #219 |

## 5. File inventory

**New (~50):**
- `src/AiDotNet.Tensors/Onnx/` root
- `Onnx/Parser/` — ProtoParser, GraphParser, InitializerParser, ValueInfoParser, OpSetVersioning (~10)
- `Onnx/OpTranslators/` — one file per ONNX op (~40 translators)
- `Onnx/OnnxImportOptions.cs`, `Onnx/OnnxImportResult.cs`, `Onnx/OnnxImporter.cs` (~3)
- `Onnx/Polyfills/` — PolyfillRegistry + composed-op polyfills for rare ops (~5)

**Modified (~5):** `CompiledInferencePlan.cs` (import-friendly constructor), `ParameterBuffer.cs` (from-onnx-initializers factory), `LazyTensorScope.cs` (import hook), `OpRegistry.cs`.

**Tests (~25):** protobuf parser tests, per-op translation correctness, ResNet-50/BERT/ViT load + reference logits match, unsupported-op reporting, Phase 2 quantized roundtrip, LSTM/GRU parity.

**Benchmarks (~4):** `OnnxImportLatencyBenchmarks` (ResNet-50/BERT/ViT), `OnnxExecutionBenchmarks` (vs ONNX Runtime).

## 6. Acceptance gate

1. Imports ResNet-50 + BERT-base + ViT-B/16 from ONNX model zoo.
2. Numerical accuracy within 1e-4 of ONNX Runtime outputs on 100 validation samples.
3. CPU latency ≥ parity with ONNX Runtime on same 3 models.
4. Unsupported operators named + reported.
5. Phase 2 QLinear ops + LSTM/GRU supported.
6. Polyfill extension point works end-to-end on a custom op.

## 7. Timeline — ~5 weeks

W1 Protobuf parser (schema, tensor initializers, value info, opset versioning).
W2 Phase 1 op translators (MatMul/Gemm/Add/Mul/activations/LayerNorm/BatchNorm).
W3 Phase 1 continued (Conv/Pool/Reshape/Transpose/Slice/Concat/Split/Gather/Attention).
W4 Phase 2 (QLinear* + LSTM/GRU).
W5 Polyfill framework + benchmarks + PR.

## 8. Dependencies

- **Blocks**: fastest-ONNX positioning as competitive story.
- **Blocked by**: #210 (Gather/Slice/Concat ops), #217 (grid_sample/interpolate/pad for some vision models), #219 (cuDNN RNN for LSTM/GRU), #218 (serialization round-trip for testing).

## 9. Risks

- **ONNX opset version drift** — pin supported opsets (11+); reject others with clear error.
- **Attention extension variants** — `com.microsoft.Attention` has ~4 variants; test each.
- **Quantized ops dequant semantics** — subtle; validate against ORT on quantized ResNet.

---

*Plan ratified 2026-04-19.*
